File: app/src/main/java/com/refrii/client/UnitActivity.java
package com.refrii.client;

import android.app.Dialog;
import android.app.DialogFragment;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.InputType;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import java.text.SimpleDateFormat;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Response;

public class UnitActivity extends AppCompatActivity {

    private Unit mUnit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_unit);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        Intent intent = getIntent();
        mUnit = (Unit) intent.getSerializableExtra("unit");

        TextView labelTextView = (TextView) findViewById(R.id.labelTextView);
        TextView stepTextView = (TextView) findViewById(R.id.stepTextView);
        TextView createdTextView = (TextView) findViewById(R.id.createdTextView);
        TextView updatedTextView = (TextView) findViewById(R.id.updatedTextView);
        ImageView labelImageView = (ImageView) findViewById(R.id.labelImageView);
        ImageView stepImageView = (ImageView) findViewById(R.id.stepImageView);
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);

        setUnit(mUnit);

        labelImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                EditLabelDialogFragment fragment = new EditLabelDialogFragment();
                fragment.setUnit(mUnit);
                fragment.show(getFragmentManager(), "edit_label");
            }
        });

        stepImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                EditStepDialogFragment fragment = new EditStepDialogFragment();
                fragment.setUnit(mUnit);
                fragment.show(getFragmentManager(), "edit_step");
            }
        });

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                updateUnit(mUnit, view);
            }
        });
    }

    public void setUnit(Unit unit) {
        TextView labelTextView = (TextView) findViewById(R.id.labelTextView);
        TextView stepTextView = (TextView) findViewById(R.id.stepTextView);
        TextView createdTextView = (TextView) findViewById(R.id.createdTextView);
        TextView updatedTextView = (TextView) findViewById(R.id.updatedTextView);

        SimpleDateFormat formatter = new SimpleDateFormat("yyyy/MM/dd HH/mm");

        labelTextView.setText(unit.getLabel());
        stepTextView.setText(String.valueOf(unit.getStep()));
        createdTextView.setText(formatter.format(unit.getCreatedAt()));
        updatedTextView.setText(formatter.format(unit.getUpdatedAt()));
    }

    public void updateUnit(Unit unit, final View view) {
        UnitService service = RetrofitFactory.getClient(UnitService.class, UnitActivity.this);
        RequestBody body = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("label", unit.getLabel())
                .addFormDataPart("step", String.valueOf(unit.getStep()))
                .build();
        Call<Unit> call = service.updateUnit(unit.getId(), body);
        call.enqueue(new BasicCallback<Unit>(UnitActivity.this) {
            @Override
            public void onResponse(Call<Unit> call, Response<Unit> response) {
                super.onResponse(call, response);

                if (response.code() == 200) {
                    Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                            .setAction("Action", null).show();
                }
            }
        });
    }

    public static class EditLabelDialogFragment extends DialogFragment {
        private EditText mEditText;
        private Unit mUnit;

        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            LayoutInflater inflater = (LayoutInflater)getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            View content = inflater.inflate(R.layout.edit_text_dialog, null);
            mEditText = content.findViewById(R.id.editText);
            mEditText.setText(mUnit.getLabel());

            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setTitle("Label")
                    .setView(content)
                    .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            String label = mEditText.getText().toString();
                            mUnit.setLabel(label);
                            ((UnitActivity) getActivity()).setUnit(mUnit);
                        }
                    })
                    .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) { }
                    });
            return builder.create();
        }

        public void setUnit(Unit unit) {
            mUnit = unit;
        }
    }

    public static class EditStepDialogFragment extends DialogFragment {
        private EditText mEditText;
        private Unit mUnit;

        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            LayoutInflater inflater = (LayoutInflater)getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            View content = inflater.inflate(R.layout.edit_text_dialog, null);
            mEditText = content.findViewById(R.id.editText);
            mEditText.setInputType(InputType.TYPE_CLASS_NUMBER);
            mEditText.setText(String.valueOf(mUnit.getStep()));

            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setTitle("Step size")
                    .setView(content)
                    .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            double step = Double.valueOf(mEditText.getText().toString());
                            mUnit.setStep(step);
                            ((UnitActivity) getActivity()).setUnit(mUnit);
                        }
                    })
                    .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) { }
                    });
            return builder.create();
        }

        public void setUnit(Unit unit) {
            mUnit = unit;
        }
    }
}

File: app/src/main/java/com/refrii/client/views/activities/UnitActivity.kt
package com.refrii.client.views.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.MenuItem
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.refrii.client.R
import com.refrii.client.RealmUtil
import com.refrii.client.models.Unit
import com.refrii.client.services.RetrofitFactory
import com.refrii.client.services.UnitService
import com.refrii.client.views.fragments.EditDoubleDialogFragment
import com.refrii.client.views.fragments.EditTextDialogFragment
import io.realm.Realm
import kotterknife.bindView
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.text.SimpleDateFormat
import java.util.*

class UnitActivity : AppCompatActivity() {

    private val toolbar: Toolbar by bindView(R.id.toolbar)
    private val labelTextView: TextView by bindView(R.id.labelTextView)
    private val stepTextView: TextView by bindView(R.id.stepTextView)
    private val createdTextView: TextView by bindView(R.id.createdTextView)
    private val updatedTextView: TextView by bindView(R.id.updatedTextView)
    private val labelImageView: ImageView by bindView(R.id.labelImageView)
    private val stepImageView: ImageView by bindView(R.id.stepImageView)
    private val fab: FloatingActionButton by bindView(R.id.fab)

    private lateinit var mRealm: Realm
    private var mUnit: Unit? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_unit)
        setSupportActionBar(toolbar)
        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.setHomeButtonEnabled(true)
        }

        Realm.init(this)
        mRealm = RealmUtil.getInstance()

        val intent = intent
        val unitId = intent.getIntExtra("unit_id", 0)

        mUnit = getUnit(unitId)
        mUnit?.let { setUnit(it) }

        onLoaded()

        labelImageView.setOnClickListener {
            mUnit?.let {
                it.label?.let {
                    val fragment = EditTextDialogFragment.newInstance("Label", it)

                    fragment.setTargetFragment(null, EDIT_LABEL_REQUEST_CODE)
                    fragment.show(fragmentManager, "edit_label")
                }
            }
        }

        stepImageView.setOnClickListener {
            mUnit?.let {
                val fragment = EditDoubleDialogFragment.newInstance("Step size", it.step)

                fragment.setTargetFragment(null, EDIT_STEP_REQUEST_CODE)
                fragment.show(fragmentManager, "edit_step")
            }
        }

        fab.setOnClickListener {
            mUnit?.let { updateUnit(it) }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode != Activity.RESULT_OK) return
        data ?: return

        when (requestCode) {
            EDIT_LABEL_REQUEST_CODE -> {
                mRealm.executeTransaction { mUnit?.label = data.getStringExtra("text") }
                mUnit?.let { setUnit(it) }
                onEdited()
            }
            EDIT_STEP_REQUEST_CODE -> {
                mRealm.executeTransaction { mUnit?.step = data.getDoubleExtra("number", 0.toDouble()) }
                mUnit?.let { setUnit(it) }
                onEdited()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        var result = true

        when (id) {
            android.R.id.home -> finish()
            else -> result = super.onOptionsItemSelected(item)
        }

        return result
    }

    private fun getUnit(id: Int): Unit? =
            mRealm.where(Unit::class.java)?.equalTo("id", id)?.findFirst()

    private fun setUnit(unit: Unit) {
        val formatter = SimpleDateFormat("yyyy/MM/dd HH/mm", Locale.getDefault())

        labelTextView.text = unit.label
        stepTextView.text = unit.step.toString()
        createdTextView.text = formatter.format(unit.createdAt)
        updatedTextView.text = formatter.format(unit.updatedAt)
    }

    private fun updateUnit(unit: Unit) {
        Log.e(TAG, "" + unit.id)
        RetrofitFactory.getClient(UnitService::class.java, this@UnitActivity)
                .updateUnit(unit.id, unit.toMultipartBody())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object: Subscriber<Unit>() {
                    override fun onCompleted() {
                        Snackbar.make(fab, "Replace with your own action", Snackbar.LENGTH_LONG)
                                .setAction("Action", null)
                                .show()
                    }

                    override fun onNext(t: Unit) {
                    }

                    override fun onError(e: Throwable) {
                        Toast.makeText(this@UnitActivity, e.message, Toast.LENGTH_LONG).show()
                    }

                })
    }

    private fun onLoaded() {
        fab.hide()
    }

    private fun onEdited() {
        fab.show()
    }

    companion object {
        private const val TAG = "UnitActivity"
        private const val EDIT_LABEL_REQUEST_CODE = 101
        private const val EDIT_STEP_REQUEST_CODE = 121
    }
}
